### DISCUSSION TITLE
YOLO NAS implementation within ROI

### DISCUSSION THREAD

**Discussion by rakshita129** [community/community#58386] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

hello sir @shaydeci i am trying to implement the yolo-nas model within a selected region of interst but due to unknown reason my code is not able to perform object detection within the frame selected . I have attached the code below , could you please help me with it .
CODE :
import torch
import cv2
from super_gradients.training import models

device = 'cuda' if torch.cuda.is_available() else 'cpu'
model = models.get("yolo_nas_s", pretrained_weights="coco").to(device)

cap = cv2.VideoCapture("C:\\Users\\admin\Downloads\\concat1.mp4")

# Check if video capture is successful
if not cap.isOpened():
    print("Failed to open video file.")
    exit()

# Get the first frame of the video
ret, frame = cap.read()

# Check if frame is empty
if not ret:
    print("Failed to read the first frame.")
    exit()

# Create a window to display the video
cv2.namedWindow('Select ROI', cv2.WINDOW_NORMAL)
frame = cv2.resize(frame, (1000, 700))
cv2.imshow('Select ROI', frame)

# Select the ROI
roi = cv2.selectROI('Select ROI', frame, fromCenter=False, showCrosshair=False)

# Close the window
cv2.destroyWindow('Select ROI')

# Set flag for object detection
detect_objects = False

# Loop through the frames and display the selected ROI
while cap.isOpened():
    ret, frame = cap.read()

    if ret:
        # Crop the frame to the selected ROI
        frame = cv2.resize(frame, (1000, 700))
        x, y, w, h = roi

        cv2.rectangle(frame, (x, y), (x+w, y+h), (0, 255, 0), 2)
        frame_roi = frame[int(y):int(y+h), int(x):int(x+w)]

        # Perform object detection when Enter is pressed
        if detect_objects:
            with torch.no_grad():
                frame_roi_tensor = torch.from_numpy(frame_roi.transpose((2, 0, 1))).float().unsqueeze(0).to(device) / 255.0
                results = model(frame_roi_tensor)

            label_names = results.names
            labels = results.xyxyn[0][:, -1].cpu().numpy()
            confidences = results.xyxyn[0][:, 4].cpu().numpy()

            for label, confidence in zip(labels, confidences):
                label_name = label_names[int(label)]
                fm = f'{label_name} {confidence:.2f}'
                x1, y1, x2, y2 = int(x + results.xyxyn[0][0, 0].item() * w), int(y + results.xyxyn[0][0, 1].item() * h), int(x + results.xyxyn[0][0, 2].item() * w), int(y + results.xyxyn[0][0, 3].item() * h)
                bgr = (0, 255, 0)
                cv2.rectangle(frame, (x1, y1), (x2, y2), bgr, 2)
                cv2.putText(frame, fm, (x1, y1), cv2.FONT_HERSHEY_SIMPLEX, 0.9, bgr, 2)

        cv2.imshow('ROI', frame)

        key = cv2.waitKey(25) & 0xFF
        if key == ord('q'):
            break
        elif key == 13:  # Enter key
            detect_objects = True

    else:
        break

# Release the video capture and close the window
cap.release()
cv2.destroyAllWindows()
